Enhance `verify-setup` buffer

fmdkdd opened the pull request:
1.  Separate first checker from other checkers in verify-setup.
      
    This makes it clearer that only one checker can be run, even though
    multiple checkers may be available.

2.  Display next checkers in verify-setup when applicable.

    This makes it easier to debug checker chaining.

Here are some screenshots:

![elisp](https://user-images.githubusercontent.com/340129/41813790-e1204e6c-773c-11e8-9141-4febd799b81b.png)

The checker that will effectively run is the first, and the available checkers for the buffer are below.  Note that the first checker is not repeated in the available checkers.  You can also see the new `next checkers` verification result.

Here is what it looks like with larger chains:

![more-checkers](https://user-images.githubusercontent.com/340129/41813792-e827dd4c-773c-11e8-8ae1-3df42c681031.png)

I've added two checkers `foobar` and `foobar2` for testing.  `foobar` was added as `next-checker` to `c/c++-clang`, and `foobar2` as `next-checker` to `foobar`.  This is all explicitly visible in the screenshot.  The first checker will show the full chain, but the description of `foobar` still shows that `foobar2` comes after it.

And the last example is for a buffer without any checker:

![no-checkers](https://user-images.githubusercontent.com/340129/41813793-e9dce86c-773c-11e8-8c4a-9fb70a4af2e1.png)

cpitclaudel commented:
Thanks, the screenshot look very nice.  Should we also show the other checkers in the current chain in the first section? We often run more than one checker on a single buffer, so it seems misleading to highlight just the first one.

In the past we've also discussed clarifying 'automatically disabled' vs 'may enable': maybe we could add an 'enable now' link next to 'may enable'?  An in fact, if we do that, we could remove that line entirely and only show the 'enable now' link next to the thing that says 'automatically disabled'?

fmdkdd commented:
>  Should we also show the other checkers in the current chain in the first section? We often run more than one checker on a single buffer, so it seems misleading to highlight just the first one.

I tend to agree.  I can put them in sequence, although it might be misleading since we cannot know if all checkers will be run (depending on the value of `current-errors`).  At least that gets you a distinction between "these checkers may run in this buffer, and these other checkers won't definitely run until you change your config".

Plus `:next-checkers` can actually be multiple checkers, and only one will eventually be picked, depending on current errors .  For now I'm using `flycheck-get-next-checker-for-buffer` which picks just *one* checker from `:next-checkers`.   A quick grep tells me that we don't have any checker with multiple `:next-checkers` set, so this point may be moot.

> In the past we've also discussed clarifying 'automatically disabled' vs 'may enable': 

:+1:  This is confusing, I'll look into it.  An `enable now` is a much better solution indeed.

fmdkdd commented:
@m-cat If you want to help, can you check out this PR and report on its usability, what could be improved, etc?  I've rebased on latest master.

m-cat commented:
Sure @fmdkdd, will be glad to help once I find some time 👍

cpitclaudel commented on doc/user/syntax-checkers.rst:
```diff
@@ -274,3 +274,57 @@ installed into a local ``node_modules`` directory:
+a *checker chain*.
+
+Some checkers chains are already setup by default in Flycheck: e.g.,
+`emacs-lisp` will follow with `emacs-lisp-checkdoc`, and `python-mypy` will
```
Maybe `will be followed by`?

cpitclaudel commented on doc/user/syntax-checkers.rst:
```diff
@@ -274,3 +274,57 @@ installed into a local ``node_modules`` directory:
+
+For a given checker, several next checkers may be specified.  Flycheck will run
+the first (in order of declaration) whose error level matches (see below) and
+whose checker can be used in the current buffer.
```
`whose checker` → `which`

cpitclaudel commented on flycheck.el:
```diff
@@ -2130,11 +2130,14 @@ into the verification results."
                  'type 'help-flycheck-checker-doc
                  'help-args (list checker))
-  (when (with-current-buffer buffer (flycheck-disabled-checker-p checker))
+  (when (buffer-local-value (flycheck-disabled-checker-p checker) buffer)
```
:clap:

fmdkdd replied:
Actually, that would have worked with the `flycheck-disabled-checkers` variable, but it doesn't with the predicate :sweat:.

cpitclaudel commented on flycheck.el:
```diff
@@ -2130,11 +2130,14 @@ into the verification results."
+  (when (buffer-local-value (flycheck-disabled-checker-p checker) buffer)
     (insert (propertize " (disabled)" 'face '(bold error))))
+  (when (eq checker (buffer-local-value 'flycheck-checker buffer))
+    (insert (propertize "(explicitly selected)" 'face 'bold)))
```
Do you need a space here?

fmdkdd replied:
I do!

cpitclaudel commented on flycheck.el:
```diff
@@ -2159,10 +2162,43 @@ into the verification results."
+          ;; button face from the checker chain result.
+          (if face
+              (insert (propertize message 'face face))
+            (insert message)))
```
But when face is non-nil, won't it override the existing face? Maybe you want `font-lock-append-text-property`?

cpitclaudel replied:
Oh I think I understand now.  Just a small change then:
`(insert (if face (propertize message …) message))` (no need to repeat the insert)

cpitclaudel commented on flycheck.el:
```diff
@@ -2159,10 +2162,43 @@ into the verification results."
+  (let ((chain))
+    (while (setq checker (flycheck-get-next-checker-for-buffer checker))
+      (push checker chain))
+    (nreverse chain)))
```
Do we only want the usable ones here? Wouldn't it better instead to include also the ones that may not be used for this buffer? (`-get-next-checker-for-buffer` calls `flycheck-may-use-next-checker`, which checks the current max error level, etc)

fmdkdd replied:
Hmm the problem with the static view of available next checkers is that it is a tree, not a list.

We could display the immediate next checkers of the current one; that is easy and not ambiguous.

cpitclaudel replied:
Indeed :thinking:
I wonder if the simplest wouldn't be to just show the list of next checkers for each checker.

cpitclaudel commented on flycheck.el:
```diff
@@ -2159,10 +2162,43 @@ into the verification results."
+                                         'help-args (list checker))
+                     (insert " "))
+                   (buffer-string)))
+                ", ")))))
```
Cool trick: `make-text-button` can take a string and propertize it; hence this should work:

```elisp
 (string-join
  (seq-map (lambda (checker)
             (make-text-button (symbol-name checker) nil
                               'type 'help-flycheck-checker-doc
                               'help-args (list checker)))
           chain)
  ", ")
```

cpitclaudel commented on flycheck.el:
```diff
@@ -2179,7 +2215,7 @@ buffer."
                    'type 'help-function
                    'help-args (list mode)))
-  (princ ":\n\n"))
+  (princ ".\n\n"))
```
?

cpitclaudel commented on flycheck.el:
```diff
@@ -2245,34 +2281,46 @@ Display a new buffer listing all syntax checkers that could be
+  ;; verification
   (when (and (buffer-file-name) (buffer-modified-p))
-    ;; Save the buffer
     (save-buffer))
```
Ugh, I didn't realize we did that.  Maybe we should do it through `save-some-buffers` instead?
I'd be happy with just a FIXME; no need to fix it now.

fmdkdd replied:
Added a FIXME.  `save-some-buffers` would let the user choose, but then we would have a few `:predicate` failing due to `flycheck-buffer-saved-p`.  Ideally we would annotate these in the verify buffer, but how do we know that `flycheck-buffer-saved-p` is used in the predicate?  Macro magic in `define-checker`?

cpitclaudel replied:
I think we can't know; the best we could do would be to show a warning if the current buffer is unsaved.

cpitclaudel commented on flycheck.el:
```diff
@@ -2245,34 +2281,46 @@ Display a new buffer listing all syntax checkers that could be
+
+        ;; First display the checkers that will run for this buffer
+        (when checker-chain
+          (progn
```
No need for progn I think

cpitclaudel commented on flycheck.el:
```diff
@@ -2245,34 +2281,46 @@ Display a new buffer listing all syntax checkers that could be
+
+        ;; Then display other checkers
+        (when other-checkers
+          (progn
```
Same here

cpitclaudel commented on flycheck.el:
```diff
@@ -2245,34 +2281,46 @@ Display a new buffer listing all syntax checkers that could be
+              (flycheck--verify-princ-checker checker buffer))))
+
+        ;; If we have no checkers at all, that's worth mentioning
+        (when (and (not checker-chain) (not other-checkers))
```
`(unless (or checker-chain other-checkers)`?

fmdkdd replied:
I always have trouble thinking about `unless` :)

fmdkdd commented:
@cpitclaudel I made some changes.  Now we display:

1. The first checker to run.   I added back the message when there isn't any first checker.
2. All the checkers that are part of the first checker's chain.  That is a *static* view, not using `-get-next-checker`, so we disregard the current error level.
3. All checkers that may be used in this buffer, but are not part of the current chain.  I added a `select` button on these to easily toggle from the verify setup buffer.
4. Any remaining (misconfigured) checker for the major mode.

cpitclaudel commented on doc/user/flycheck-versus-flymake.rst:
```diff
@@ -225,9 +225,8 @@ run in sequence rather than concurrently.  For instance, Flycheck will check PHP
-least one error, then the subsequent checkers would not be run (by default; this
-behavior is configurable).
+all of these tools in the buffer.  This behavior is fully configurable; see
+:ref:`flycheck-checker-chains`.
```
I think the 'this behavior' part is unclear with the previous sentence removed; don't you think?

fmdkdd replied:
I agree.  I'm not sure why I removed it in the first place.

cpitclaudel commented on flycheck.el:
```diff
@@ -2173,10 +2200,45 @@ into the verification results."
+Return the transitive closure of the next-checker relation.  The
+return value is a list of checkers, not including CHECKER."
+  (let ((next (flycheck-get-next-checkers checker)))
+    (seq-uniq (append next (seq-mapcat #'flycheck-all-next-checkers next)))))
```
This is going to infloop if someone adds a checker as its own next-checker, right? I think our verification code will likely infloop as well, though, so this is not worse

fmdkdd replied:
Yes, this assumes we don't have cycles.  Ideally we want to ensure there are no cycles by construction, but I'm not sure how to do that in `define-checker` and when the `:next-checkers` property is altered after.

Do you want me to harden this function to ensure we don't revisit checkers twice?

fmdkdd replied:
@cpitclaudel I've changed the code to avoid visiting checkers twice.  Can you look?

cpitclaudel commented on flycheck.el:
```diff
@@ -2173,10 +2200,45 @@ into the verification results."
+    (list
+     (flycheck-verification-result-new
+      :label "next checkers"
+      ;; We use `insert-text-button' to preserve the button properties in
```
That comment needs to be updated :)

cpitclaudel commented on flycheck.el:
```diff
@@ -2173,10 +2200,45 @@ into the verification results."
+                                             'type 'help-flycheck-checker-doc
+                                             'help-args (list checker)))
+                         next)
+                ", ")))))
```
Not that I look at this again, I think this is an ideal candidate for mapconcat :)

cpitclaudel commented:
Thanks for the update. I just gave it a try. It looks great.
One thing that I noticed: in my Python buffer, I have `flycheck-disabled-checkers` set to `(python-flake8 python-mypy)`; in the old code I get `(disabled)` next to these checkers, but in the new version they don't have this mark any more.  Is that intended?

fmdkdd commented:
>  I get (disabled) next to these checkers, but in the new version they don't have this mark any more. Is that intended?

Maybe you are using an old version of the PR?  I did have that bug, but I fixed it.

cpitclaudel commented:
> Maybe you are using an old version of the PR? I did have that bug, but I fixed it.

Yes, that seems to be it!  It looks gorgeous :)
Please merge :rocket: Don't forget the changelog entry!
